### ISSUE TITLE
Recursion Error in the function traverseUp in util.py on windows.

### ISSUE THREAD

**Issue by noc7c9** [autozimu/LanguageClient-neovim#93]
![untitled](https://user-images.githubusercontent.com/3779513/29249456-0794afb4-8073-11e7-8900-cfcb321ebba5.png)

I did some logging of the function:
```py
def traverseUp(folder: str, stop) -> str:
    print(folder, file=fp)
    fp.flush()
    if stop(folder):
        return folder
    elif folder == "/" or folder == "":
        return None
    else:
        return traverseUp(os.path.dirname(folder), stop)
```

```
c:\users\user\downloads\test.py
c:\users\user\downloads
c:\users\user
c:\users
c:\
c:\
c:\
c:\
c:\
c:\
c:\
...
```
And so on.

This happened to me with the python language server on a standalone script but given the nature of the issue I doubt that matters. Specifically it'll happen any time there is no clear root path.

```diff
-    elif folder == "/" or folder == "":
+    elif folder == "/" or folder == "" or folder == os.path.dirname(folder):
```
That fixes it, sorry I'm too tired for a proper pull request. xD

**Comment by noc7c9** [comment]
This is unrelated but I noticed that the `haskell` `rootPath` detection has two calls of `traverseUp`.
```py
        rootPath = (traverseUp(
                        filepath,
                        lambda folder: 
                            os.path.exists(os.path.join(folder, "stack.yaml"))) or 
                    traverseUp( 
                        filepath,
                        lambda folder: 
                            os.path.exists(os.path.join(folder, ".cabal"))))
```

Shouldn't it be:
```py
        rootPath = traverseUp(
                        filepath,
                        lambda folder: (
                            os.path.exists(os.path.join(folder, "stack.yaml")) or
                            os.path.exists(os.path.join(folder, ".cabal"))))
```

I guess it doesn't really matter, its probably negligibly faster.

**Comment by gleachkr** [comment]
That logic is actually intended. A `stack.yaml` in the path should supercede a `.cabal`.

**Comment by autozimu** [comment]
Fixed by 1d2b0cb7875415bbfa27474bca7edf3ef23a661d
